src/mol-model/structure/structure/mapping.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
+ *
+ * @author Alexander Rose <alexander.rose@weirdbyte.de>
+ */
+
+import { OrderedSet } from '../../../mol-data/int/ordered-set';
+import { ElementIndex } from '../model/indexing';
+import { Structure } from './structure';
+import { Unit } from './unit';
+
+/** Serial index of an element in the structure across all units */
+export type SerialIndex = { readonly '@type': 'serial-index' } & number
+
+export interface SerialMapping {
+    /** Cumulative count of preceding elements for each unit */
+    cumulativeUnitElementCount: ArrayLike<number>
+    /** Unit index for each serial element in the structure */
+    unitIndices: ArrayLike<number>
+    /** Element index for each serial element in the structure */
+    elementIndices: ArrayLike<ElementIndex>
+    /** Get serial index of element in the structure */
+    getSerialIndex: (unit: Unit, element: ElementIndex) => SerialIndex
+}
+
+export function getSerialMapping(structure: Structure): SerialMapping {
+    const { units, elementCount, unitIndexMap } = structure;
+    const cumulativeUnitElementCount = new Uint32Array(units.length);
+    const unitIndices = new Uint32Array(elementCount);
+    const elementIndices = new Uint32Array(elementCount) as unknown as ElementIndex[];
+    for (let i = 0, m = 0, il = units.length; i < il; ++i) {
+        cumulativeUnitElementCount[i] = m;
+        const { elements } = units[i];
+        for (let j = 0, jl = elements.length; j < jl; ++j) {
+            const mj = m + j;
+            unitIndices[mj] = i;
+            elementIndices[mj] = elements[j];
+        }
+        m += elements.length;
+    }
+    return {
+        cumulativeUnitElementCount,
+        unitIndices,
+        elementIndices,
+
+        getSerialIndex: (unit, element) => cumulativeUnitElementCount[unitIndexMap.get(unit.id)] + OrderedSet.indexOf(unit.elements, element) as SerialIndex
+    };
+}
+
+//
+
+export interface IntraUnitBondMapping {
+    bondCount: number
+    unitIndex: ArrayLike<number>
+    unitEdgeIndex: ArrayLike<number>
+    unitGroupIndex: ArrayLike<number>
+}
```

Comment:
Instead of this, would it make sense to refactor the `LinkBuilderProps` (or add `StructureLinkBuilderProps`) with this pattern:

```ts
interface StrutureLinkBuilderProps {
  callback(unitIndex: number, unitGroupIndex: number, unitEdgeIndex: number): T {
  }
}
```